### PR TITLE
Correlation rename fixes

### PR DESCRIPTION
**Pull request by julien-thierry**
As mentioned on https://github.com/dynup/kpatch/pull/1043.

This change actually breaks some tests I was rebasing for rhel-8.0 with the following kind of error:
(this is on gcc-static-local-var-4.patch)
ERROR: aio.o: kpatch_create_patches_sections: 2708: lookup_local_symbol put_aio_ring_file.isra.19

The issue is due to the fact that we dropped symbol renaming when correlating. For symbols that end with status "SAME" there is no problem as we don't include those in the final object. For symbols that have status "CHANGED", we need to lookup the original symbols to know their address/position in the kernel/module getting patched, but at the point of the lookup, we only have the name from the patched object.
In the above example, put_aio_ring_file.isra.19 from the patched object gets correlated to put_aio_ring_file.isra.20 from the original object. But when creating the patch module itself, we no longer have the information of the original name.

My suggestion is to always rename mangled elements with the name of the object they get correlated to (last patch).

I also took the opportunity to do some cleanup related to correlation.

### CODE REVIEW

**Comment by julien-thierry** [comment]
I added a unit test highlighting the issue:
https://github.com/dynup/kpatch-unit-test-objs/pull/19

File: kpatch-build/kpatch-elf.c [hunk 1]
```diff
@@ -858,6 +860,8 @@ void kpatch_elf_teardown(struct kpatch_elf *kelf)
 	}
 
 	list_for_each_entry_safe(sym, safesym, &kelf->symbols, list) {
+		if (sym->twin)
+			sym->twin = NULL;
```
Comment (joe-lawrence) [inline kpatch-build/kpatch-elf.c]: Just curious, where do the stale references occur?  
In this instance, wouldn't the subsequent memset be clearing sym->twin value anyway?

File: kpatch-build/create-diff-object.c [hunk 2]
```diff
@@ -760,11 +760,39 @@ static void kpatch_compare_symbols(struct list_head *symlist)
 	}
 }
 
+#define CORRELATE_ELEMENTS(e1, e2)				\
```
Comment (joe-lawrence) [inline kpatch-build/create-diff-object.c]: nit: not a big deal, but wouldn't the compiler figure out if inlining is a win here?

File: kpatch-build/kpatch-elf.c (hunk 1 above)
Comment (julien-thierry) [reply kpatch-build/kpatch-elf.c]: Whoops, typo from my end, this should be like for the sections:
`sym->twin->twin = NULL;`

I don't think there is currently a bug due to that. It is just to make sure no-one is tempted to check the `twin` pointers of the patched object when building the final object as they reference freed object.

File: kpatch-build/create-diff-object.c (hunk 2 above)
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: I didn't go for a macro for inlining here but because I can use the macro for both `struct symbol *` and `struct section *` expressions.

If that's not considered good practice, I can have two static functions, but their code will be pretty much identical.

File: kpatch-build/create-diff-object.c [hunk 3]
```diff
@@ -760,11 +760,45 @@ static void kpatch_compare_symbols(struct list_head *symlist)
...
+do {								\
+	e1->twin = e2;						\
+	e2->twin = e1;						\
+	/* set initial status, might change */			\
+	e1->status = e2->status = SAME;				\
+	if (strcmp(e1->name, e2->name)) {			\
+		/* Rename mangled element */			\
+		log_debug("renaming %s to %s\n",		\
```
Comment (sm00th) [inline kpatch-build/create-diff-object.c]: Would be nice if we could somehow differentiate between sections and symbols here.

File: kpatch-build/create-diff-object.c [hunk 4]
```diff
@@ -766,6 +766,12 @@ do {								\
 	e2->twin = e1;						\
 	/* set initial status, might change */			\
 	e1->status = e2->status = SAME;				\
+	if (strcmp(e1->name, e2->name)) {			\
+		/* Rename mangled element */			\
+		log_debug("renaming %s to %s\n",		\
+			  e2->name, e1->name);			\
+		e2->name = strdup(e1->name);			\
```
[the remark this answers is not in the document]
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: The name of an element might point to a string in a elf file which might be closed when we still are using e2, so we need to dup (or not close the original elf nor the patched elf until we quit the program).

As for memory leaking, I must admit I followed the way the code Josh deleted used to handle things. The issue is that at this point a section or symbol name could point to memory mapped from an elf file or memory allocated through malloc.
Since kpatch is just a "one-off" program, having this memory allocated until the end of the program should be too much of a big deal. But if we want to avoid that we can strdup all names and free them in the kpatch_elf_teardown().

Let me know if we shoudl go with that.

File: kpatch-build/create-diff-object.c (hunk 3 above)
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: So, have two separate static functions?

File: kpatch-build/create-diff-object.c [hunk 5]
```diff
@@ -760,11 +760,39 @@ static void kpatch_compare_symbols(struct list_head *symlist)
...
+	if (sec->sym)
+		CORRELATE_ELEMENTS(sec->sym, sec->twin->sym);
+}
+
 static void kpatch_correlate_sections(struct list_head *seclist1, struct list_head *seclist2)
 {
 	struct section *sec1, *sec2;
 
 	list_for_each_entry(sec1, seclist1, list) {
+		if (sec1->twin)
```
[the remark this answers is not in the document]
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: The `kpatch_process_section()` might already have done the correlation for an associated rela/base section. In which case, looking for another element with a matching mangled name can match to a different object (since the correct one will be already correlated, i.e. `sec2->twin != NULL`, we can't match to it again).

File: kpatch-build/create-diff-object.c [hunk 6]
```diff
@@ -800,6 +828,8 @@ static void kpatch_correlate_symbols(struct list_head *symlist1, struct list_hea
 	struct symbol *sym1, *sym2;
 
 	list_for_each_entry(sym1, symlist1, list) {
+		if (sym1->twin)
```
[the remark this answers is not in the document]
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: More or less the same story as above. `kpatch_process_section()` perform the correlation for the section symbol or a symbol that is bundled.

File: kpatch-build/create-diff-object.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (sm00th) [reply kpatch-build/create-diff-object.c]: Well, I personally try to keep leaks to minimum even for "one-offs", but I did see a comment somewhere in the same vein of "we don't care since its a one-off", so I'll leave it for @jpoimboe to decide.

File: kpatch-build/create-diff-object.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (sm00th) [reply kpatch-build/create-diff-object.c]: Do you mean `kpatch_correlate_sections()`? I don't see how that would happen since it goes through each list only once.

File: kpatch-build/create-diff-object.c (hunk 3 above)
Comment (sm00th) [reply kpatch-build/create-diff-object.c]: Well every way of doing this seems to be as inconvenient as separate functions... But since those are bundled together maybe a `log_debug` at the beginning of `kpatch_correlate_symbols` and `kpatch_correlate_sections` will be enough.

File: kpatch-build/create-diff-object.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: No, sorry I meant, `kpatch_process_section_correlation()` which I introduce in this commit.

Maybe its introduction is being a bit overzealous, but taking multiple objects  with matching mangled names into consideration (which was the original motivation for @jpoimboe 's change), I don't know how e.g. we guarantee we would avoid the following scenario:
orig:
- sym_a -> bundled into sec_a

patched:
- sym_b -> bundled into sec_b
- sym_c -> bundled into sec_c

If both sym_b and sym_c have names matching sym_a through mangling, we could end up with sym_a being correlated with sym_b but sec_a being correlated with sec_c.
Same thing with rela/base sections.

So basically I introduced the `kpatch_process_section_correlation()` to avoid that. But the consequence is that when you correlate a section, you might correlate another one (and a symbol) before having treated in in the list_foreach. If that happens, it needs to be skipped.

File: kpatch-build/create-diff-object.c (hunk 3 above)
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: Oh, so you meant differentiate the message printed?

Yes, that makes sense, I'll do that in the next version.

File: kpatch-build/create-diff-object.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: I could only see two ways to solve this:
- Always use malloc for the name of symbols and sections, but that would make us always use more memory than we need to avoid a leak that the OS will cleanup once the program finishes.
- Have a way to distinguish whether the name was allocated or not and check at the end.

First solution doesn't seems a bit wasteful just to avoid a leak that isn't that much of an issue (in my opinion) and I would prefer to keep the leak.
Second solution can be a bit awkward, but after sleeping on it we could pretend to make the distinction a bit more "useful". We could have a field `orig_name` in section and symbol structures, which by default would stay `NULL`. And if an element gets renamed `orig_name` would point to the string in the ELF while `name` would point to freeable memory. So we can know there is a need to free `name` if `orig_name != NULL` or something like that. While there is no big need for keeping the original name, I can at least see some use to it for debugging (e.g. could access it through GDB) or logging purposes.

Otherwise we can also keep the leak.

Let me know which you prefer.

[the remark this answers is not in the document]
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: Also, I just check, we already have the same leak when renaming static local variables.

File: kpatch-build/create-diff-object.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (sm00th) [reply kpatch-build/create-diff-object.c]: Ok, that makes more sense now. Not sure how correct is this approach but I don't have a better solution on hand.

File: kpatch-build/create-diff-object.c (hunk 2 above)
Comment (joe-lawrence) [reply kpatch-build/create-diff-object.c]: Ah, I see.  I don't have a strong preference either way then.

File: kpatch-build/create-diff-object.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (joe-lawrence) [reply kpatch-build/create-diff-object.c]: FWIW, some time in the past I wrapped kpatch-build with valgrind and tried to get the non-paths to be mostly silent.  For a userspace program, I wasn't going to pull my hair out getting all the various error paths to be perfect... but at the very least a clean successful build is helpful to find any new bad behavior.  /two-cents

**Comment by julien-thierry** [comment]
I updated according to the comments.

I also added some RFC commits that are not needed for the functionality but that prevent the leaks related to renaming.

File: kpatch-build/create-diff-object.c [hunk 7]
```diff
@@ -760,11 +760,72 @@ static void kpatch_compare_symbols(struct list_head *symlist)
...
+
+static void make_correlated_sections(struct section *sec1, struct section *sec2)
+{
+	CORRELATE_ELEMENTS(sec1, sec2, "section");
+}
+
+static void make_correlated_symbols(struct symbol *sym1, struct symbol *sym2)
+{
+	CORRELATE_ELEMENTS(sym1, sym2, "symbol");
+}
```
Comment (jpoimboe) [inline kpatch-build/create-diff-object.c]: Naming suggestions: make_correlated_sections() almost makes it sound like it's creating a section - how about renaming to kpatch_correlate_section()?   That also gives it a name analagous to its caller kpatch_correlate_sections().  And a similar comment for make_correlated_symbols().

And that would also imply a CORRELATE_ELEMENTS -> CORRELATE_ELEMENT rename for consistency - yes, it's correlating two elements, but depending on one's perspective they might be considered the same element :-)

File: kpatch-build/create-diff-object.c [hunk 8]
```diff
@@ -1063,12 +1117,7 @@ static void kpatch_correlate_static_local_variables(struct kpatch_elf *base,
 				ERROR("sections %s and %s aren't correlated for symbol %s",
 				      sym->sec->name, patched_sym->sec->name, sym->name);
 
-			log_debug("renaming and correlating static local %s to %s\n",
-				  patched_sym->name, sym->name);
-
-			patched_sym->name = strdup(sym->name);
-			sym->twin = patched_sym;
-			patched_sym->twin = sym;
+			CORRELATE_ELEMENTS(sym, patched_sym, "static local");
```
Comment (jpoimboe) [inline kpatch-build/create-diff-object.c]: Here you might as well call a kpatch_correlate_static_local_variable() wrapper which hides the ugly macro, similar to the other two helpers.

File: kpatch-build/create-diff-object.c [hunk 9]
```diff
@@ -1063,12 +1117,7 @@ static void kpatch_correlate_static_local_variables(struct kpatch_elf *base,
...
 
 			/* set initial status, might change */
 			sym->status = patched_sym->status = SAME;
```
Comment (jpoimboe) [inline kpatch-build/create-diff-object.c]: I think this is no longer needed since CORRELATE_ELEMENTS does this?

File: kpatch-build/create-diff-object.c (hunk 7 above)
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: I was a bit worried about the potential kpatch_correlate_section/sections confusion.
But now that you mention it, I agree it makes more sense the way you suggest.
I'll do the upate.

File: kpatch-build/create-diff-object.c (hunk 8 above)
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: Will do.

File: kpatch-build/create-diff-object.c (hunk 9 above)
Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: Yes, well spotted!

File: kpatch-build/create-diff-object.c [hunk 10]
```diff
@@ -1063,27 +1114,14 @@ static void kpatch_correlate_static_local_variables(struct kpatch_elf *base,
...
 
-				sym->sec->secsym->twin = patched_sym->sec->secsym;
-				patched_sym->sec->secsym->twin = sym->sec->secsym;
-
-				if (sym->sec->rela && patched_sym->sec->rela) {
-					sym->sec->rela->twin = patched_sym->sec->rela;
-					patched_sym->sec->rela->twin = sym->sec->rela;
-				}
+				if (sym->sec->rela && patched_sym->sec->rela)
+					kpatch_correlate_section(sym->sec->rela, patched_sym->sec->rela);
```
Comment (jpoimboe) [inline kpatch-build/create-diff-object.c]: This is all quite similar to what kpatch_process_section_correlation() does... could it all be replaced with
```
if (bundled) {
        kpatch_correlate_section(sym->sec, patched_sym->sec);
        kpatch_process_section_correlation(sym->sec);
}
```
?

And for that matter I wonder if it would be better to have kpatch_correlate_section() automatically do all the work in kpatch_process_section_correlation()?  Then kpatch_correlate_section() could call a lower-level __kpatch_correlate_section() version which doesn't do that extra work.  So kpatch_correlate_section() would be a higher-level interface which just does the right thing, and then this could just be
```
if (bundled)
        kpatch_correlate_section(sym->sec, patched_sym->sec);
```
And similarly kpatch_correlate_sections() wouldn't need to call kpatch_process_section_correlation().

Comment (julien-thierry) [reply kpatch-build/create-diff-object.c]: Yes, having kpatch_correlate_section() do all the needed work seems like a good suggestion.
Thanks.

**Comment by jpoimboe** [comment]
@julien-thierry once the test PR is merged, you can rebase this PR to update the test/unit/objs ref to reference the new test PR.  Then feel free to merge this one.
